src/GraphQL/Subscription/ObservableExtensions.cs

```diff
@@ -0,0 +1,252 @@
+                    // events queued will start ProcessAllEventsInQueueAsync
+                }
+            }
+
+            /// <summary>
+            /// Wait for the transform to complete and push the data (or error) back to the observer.
+            /// If the observer has been disposed, then data and errors are ignored.
+            /// </summary>
+            private async Task ProcessDataAsync(ValueTask<TOut> dataTask)
+            {
+                if (_token.IsCancellationRequested)
+                    return;
+                TOut dataOut;
+                try
+                {
+                    dataOut = await dataTask.ConfigureAwait(false);
+                }
+                catch (Exception error)
+                {
+                    if (!_token.IsCancellationRequested)
+                        _observer.OnError(error);
+                    return;
+                }
+                if (!_token.IsCancellationRequested)
+                    _observer.OnNext(dataOut);
+            }
+
+            /// <summary>
+            /// Wait for the transform to complete and push the error back to the observer.
+            /// If the observer has been disposed, then errors are ignored.
+            /// </summary>
+            private async Task ProcessErrorAsync(ValueTask<Exception> errorTask)
+            {
+                if (_token.IsCancellationRequested)
+                    return;
+                Exception errorOut;
+                try
+                {
+                    errorOut = await errorTask.ConfigureAwait(false);
+                }
+                catch (Exception error)
+                {
+                    if (!_token.IsCancellationRequested)
+                        _observer.OnError(error);
+                    return;
+                }
+                if (!_token.IsCancellationRequested)
+                    _observer.OnError(errorOut);
+            }
+
+            /// <summary>
+            /// Push a completion notice back to the observer.
+            /// If the observer has been disposed, ignore.
+            /// </summary>
+            private void ProcessCompletion()
+            {
+                if (!_token.IsCancellationRequested)
+                    _observer.OnCompleted();
+            }
+
+            /// <summary>
+            /// Disposes of the underlying observable sequence
+            /// </summary>
+            public void Dispose()
+            {
+                var cts = Interlocked.Exchange(ref _cancellationTokenSource, null);
+                if (cts == null)
+                    return;
+                // cancel pending operations and prevent pending operations
+                // from returning data after the observable has been detached
+                cts.Cancel();
+                // dispose the cancellation token source
+                cts.Dispose();
+                // detach the observable sequence
+                _disposeAction?.Invoke();
```

Comment:
What should go first - token cancellation or subscription disposing?

Reply:
I think this way is best.  In this particular scenario, I am not sure it truly matters much.  But there is no callbacks for the cancellation token so it will execute immediately, preventing further notifications.  Then it attempts to disconnect the subscription.

Reply:
It would matter if we were used `ThrowIfCancellationRequested` and wanted to unsubscribe gracefully without (or as least as possible) a bunch of `OperationCancelled` exceptions - first unsubscribe, then dispose cancellation token source.

Reply:
I still think this way is better.  However, I will move the short-circuit prior to the transformation.